[Dynamic Instrumentation] Symbols Upload

GreenMatan opened the pull request:
<!--
* Contributors are encouraged to read our [CONTRIBUTING](/CONTRIBUTING.md) documentation.
* Both Contributor and Reviewer Checklists are available at https://datadoghq.dev/datadog-agent/guidelines/contributing/#pull-requests.
* The pull request:
  * Should only fix one issue or add one feature at a time.
  * Must update the test suite for the relevant functionality.
  * Should pass all status checks before being reviewed or merged.
* Commit titles should be prefixed with general area of pull request's change.
* Please fill the below sections if possible with relevant information or links.
-->
### What does this PR do?
Completes the work started in #39020 - utilizing symbols extraction functionality and uploads it to debugger's backend for tracked Go processes.

The debugger backend delivers symdb upload request through RC. When a service receives this request, the system-probe reads it's DWARF and start extracting symbols in hierarchy of Package -> Function -> Args/Locals.
The symbols are uploaded with an iterator to avoid excessive memory usage during the upload.  Currently, we cap the upload batches to packages accumulating to >= 10k functions.

### Motivation
Better experience when creating probes through DI UI. With symbols users are able to simply select which function they want to probe without Source code integration and without intimate knowledge of the way Go constructs functions fully qualified names.

GreenMatan committed:
dyninst: implement symbols upload

andreimatei committed:
dyninst: improve batching of symdb uploads

Push the generation of envelopes and metadata into the symdbSender,
below the chunking of data, so that we don't need to marshal and
unmarshal repeatedly to deal with these wrappers.
A SymDBUploader is now tied to one upload (i.e. one binary).

GreenMatan committed:
changed the lang to go + root scope to package

GreenMatan committed:
fixed malformed json

GreenMatan committed:
added GoQualifiedName into LanguageSpecifics

andreimatei committed:
dyninst: don't use the batcher for uploads

Move away from using the Batcher for uploading SymDB. The Batcher was
written for logs, which are pretty different -- they're continuous in
nature and so the main point of the Batcher is to flush on a timer. We
don't need the timer behavior for SymDB upload, and so we were gaining
very little for using it for SymDB; replacing it is only a few lines of
code. Sharing the logic came with a bunch of issues:
- we had multiple concurrency levels, making it more difficult to reason
  about the code: requestUpload was spawning a goroutine and the also
  the batcher was spawning goroutines for each upload batch
- because of the async nature, the Batcher allowed for out of order
  uploads. This will be a problem if we start uploading "amend" batches,
  which rely on some scopes already existing in the database
- the error behavior is funky -- the interface currently doesn't return
  any upload errors (although I had a patch to store the first error and
  return it on Stop())
- (this is the big one) the fact that uploads were async, decoupled from
  enqueuing the data, makes the memory usage unpredictable -- there's no
  control over how many uploads are in flight, which can mean that the
  whole symdb data is effectively in memory at once. Instead, we want slow
  uploads to backpressure the dwarf exploration.

We could make the batcher more configurable to address some of these,
but it's not worth it; the SymDB code is better of not depending on it
and doing its own batching and synchronous uploading.

andreimatei reviewed (commented):
I pushed a commit removing the dependency on the `Batcher` since I had issues with it; hope you don't mind. Please take a look.

andreimatei committed:
dyninst: add integration tests for SymDB upload

andreimatei commented on pkg/dyninst/procmon/analyze.go:
```diff
@@ -131,7 +131,7 @@ func analyzeProcess(
-	if ddEnv.serviceName == "" || !ddEnv.diEnabled {
+	if ddEnv.serviceName == "" || ddEnv.serviceVersion == "" || ddEnv.environmentName == "" || !ddEnv.diEnabled {
 		log.Tracef(
 			"process %d is not interesting: service name is %q, %s=%t",
```
let's include all the relevant env vars in here

andreimatei commented on pkg/dyninst/procmon/analyze.go:
```diff
@@ -131,7 +131,7 @@ func analyzeProcess(
 		)
 	}
-	if ddEnv.serviceName == "" || !ddEnv.diEnabled {
+	if ddEnv.serviceName == "" || ddEnv.serviceVersion == "" || ddEnv.environmentName == "" || !ddEnv.diEnabled {
```
Let's make sure we're all on the same page here -- that we need all three of service name, env, and version in order to monitor a process. All of these are indeed needed for that process to be targeted for probing through RC?
From the UI, I seem to be able to create probes without filtering on either env or version, AFAICS.

cc @ajwerner @piob-io

GreenMatan replied:
yes, all three are required in all other languages, as [described in our docs](https://docs.datadoghq.com/tracing/dynamic_instrumentation/#:~:text=Unified%20Service%20Tagging%20tags%20service%2C%20env%2C%20and%20version%20are%20applied%20to%20your%20deployment.). I think it's safe to require them. Also, in the upcoming 'on by default' efforts, we'd like to target preprod envs which will require anyhow to set the environment.
WDYT?

GreenMatan replied:
Keeping this conv open until it's fully tested

andreimatei committed:
dyninst: improve SymDB upload control

- upload symbols for one process at a time, to bound resource usage
- handle the remote config upload request flapping between true/false ->
stop an upload if it's in progress when we get an update saying that
we shouldn't upload data any more, and allow re-uploading later if we're
again asked for data for a process
- support stopping the symdbManager on module shutdown

andreimatei reviewed (approved):
I massaged all the commits so that they all compile and pass tests.
I'm approving this now; I think it's good to go.
Do a pass and give your stamp when you can.

GreenMatan committed:
changed scopeType of go funcs wihthout receiver from  ->

GreenMatan commented on pkg/dyninst/module/state.go:
```diff
@@ -142,3 +145,16 @@ func (ps *processStore) getRuntimeID(procID actuator.ProcessID) (procRuntimeID,
+		return false // Process gone or already started
+	}
+
+	proc.symdbUploadStarted = true
```
This conv is no longer relevant, you already implemented aborting in flight symdb uploads
Resolving

andreimatei committed:
dyninst/symdb: properly synchronize uploads with RC state

Before this patch, we were uploading data anew every time we got an
"update" from the Scraper telling us that SymDB upload is requested; we
didn't keep track of the previous state of that flag so we were
uploading repeatedly. We also weren't responding to process removal
notifications from the procmon.

Now, we don't upload repeatedly, and we cancel an upload when the
process is removed.

andreimatei committed:
dyninst/symdb: gate SymDB upload on env var

Introduce DD_SYMBOL_DATABASE_UPLOAD_ENABLED in the system probe (name
matches the env var used by tracers in other languages), gating the
symdb uploads. The env var defaults to false for now; the plan is to
flip the default to true once we introduce a way to control the rollout
through the backend.
